blockchain/vm/evm.go

```diff
@@ -538,6 +538,33 @@ func (evm *EVM) CreateWithAddress(caller types.ContractRef, code []byte, gas uin
 	return evm.create(caller, codeAndHash, gas, value, contractAddr, humanReadable, codeFormat)
 }
 
+func (evm *EVM) CurrentCodeFormat() params.CodeFormat {
+	switch {
+	case evm.chainRules.IsIstanbul:
+		return params.CodeFormatEVM2
+	default:
+		return params.CodeFormatEVM
+	}
+}
+
+func (evm *EVM) GetPrecompiledContract(caller, address common.Address) map[common.Address]PrecompiledContract {
+	switch {
+	case evm.chainRules.IsIstanbul:
+		// Below code is added to make vmLog(0x09), feePayer(0x0a), validateSender(0x0b) precompiled contract
+		// works after the istanbul incompatible change
+		// (1) check whether the contract(caller) is deployed before istanbul incompatible change
+		// (2) then, check whether the precompiled contract(address) is 0x09, 0x0a, 0x0b
+		// (3) if it is, use old precompiled contract set
+		if evm.StateDB.GetCodeFormat(caller) == params.CodeFormatEVM &&
+			(address == common.HexToAddress("0x09") || address == common.HexToAddress("0x0a") || address == common.HexToAddress("0x0b")) {
```

Comment:
@yoomee1313 Can this code be simplified like below?
```suggestion
		if evm.StateDB.GetCodeFormat(caller) == params.CodeFormatEVM {
```

Reply:
Gas Costs of bn256 precompiled contracts(0x06, 0x07, 0x08) will be changed in IstanbulCompatible Change. So, this code is necessary!

Reply:
@yoomee1313 Could you explain more or make the next PR about how you are going to change with gas costs? I don't get how things will change.

Reply:
As we discussed at offline, 0x06, 0x07, 0x08 should use istanbul precompiled contract map even though the contract is deployed before the istanbulCompatible change. Only 0x09, 0x0a, 0x0b needs constantinople precompiled contract map if the contract is deployed before the change.

Reply:
@yoomee1313 Thanks for explaining. I get it now :)

Reply:
I think this is not a good idea. Let's discuss this in offline together.